edb/schema/utils.py

```diff
@@ -1109,18 +1074,65 @@ def ensure_union_type(
     return schema, uniontype, created
 
 
-def get_union_type(
+def simplify_union_types(
     schema: s_schema.Schema,
-    types: Iterable[s_types.Type],
-    *,
-    opaque: bool = False,
-    module: Optional[str] = None,
-) -> Tuple[s_schema.Schema, s_types.Type]:
+    types: Sequence[s_types.Type],
+) -> Sequence[s_types.Type]:
+
+    from edb.schema import types as s_types
+
+    components: Set[s_types.Type] = set()
+    for t in types:
+        union_of = t.get_union_of(schema)
+        if union_of:
+            components.update(union_of.objects(schema))
+        else:
+            components.add(t)
+
+    if all(isinstance(c, s_types.InheritingType) for c in components):
+        return list(minimize_class_set_by_most_generic(
+            schema,
+            cast(Set[s_types.InheritingType], components),
+        ))
+    else:
+        return list(components)
 
-    schema, union, _ = ensure_union_type(
-        schema, types, opaque=opaque, module=module)
 
-    return schema, union
+def simplify_union_types_preserve_derived(
+    schema: s_schema.Schema,
+    types: Sequence[s_types.Type],
+) -> Sequence[s_types.Type]:
```

Comment:
I probably would have kept the `preserve_derived` flag and made this one function, but I also don't object to having it be separated.

Could you add docstring comments to both?

Reply:
I couldn't figure out a good way to do it, but some day it would be nice to have the preserve derived variant removed entirely.

Reply:
The code that it was lifted from was conditional though, right?

Reply:
Yeah, but I think ultimately it may be best to figure out how to move the special logic all the way out to `func.compile_operator`, which is the only place which uses `preserve_derived`